### PR TITLE
fix(relocation): Change task scheduling to always be last action

### PR DESCRIPTION
**Pull request by azaslavsky**
This ensures that the current task is marked "done" in the database before the next task is kicked off. Since this is handled by the `retry_or_fail_relocation` context manager, all next task calls must be moved outside of said context manager, becoming the very last procedures completed in their owning task.

Because this change required genericizing over the "schedule the next task" operation, all such operations have been converted in `.apply_async()` calls, and a new `NextTask` class has been added to track task invocations that we want to kick off at the end of the current task.

Issue: getsentry/team-ospo#214

### CODE REVIEW

File: src/sentry/testutils/helpers/task_runner.py [hunk 1]
```diff
@@ -41,8 +41,8 @@ def BurstTaskRunner():
 
     job_queue = []
 
-    def apply_async(self, args=(), kwargs=(), countdown=None, queue=None):
-        job_queue.append((self, args, kwargs))
+    def apply_async(self, args=(), kwargs=None, countdown=None, queue=None):
```
Comment (hubertdeng123) [inline src/sentry/testutils/helpers/task_runner.py:44]: What's the reason behind this change?

**Review by hubertdeng123 — COMMENTED** [review]
How difficult is it to get a test case to make sure the race condition is accounted for here?

File: src/sentry/testutils/helpers/task_runner.py (hunk 1 above)
Comment (azaslavsky) [reply src/sentry/testutils/helpers/task_runner.py:44]: On line 54, we try to call `**kwargs`, which ofc won't work on a tuple. I think whoever originally wrote this code never used it with a non-default value. Specifically, I suspect that all of the code tested via this runner used `.delay()` to schedule tasks, which is a convenience wrapper around `.apply_async()` and automatically passes `{}` to its `kwargs` param. So in practice, all prior invocations ignored the incorrect default value of `()` and used `{}` under the hood.

Now that we are using this with a test where we call `.apply_async()` directly, so in cases where we don't explicitly set `kwargs`, the default actually gets invoked. So we need to make sure that in cases where no `kwargs` is passed, we use an empty dict.

**Review by hubertdeng123 — APPROVED** [review]
